pandas/tests/io/excel/test_xlsxwriter.py

```diff
@@ -61,3 +66,173 @@ def test_write_append_mode_raises(ext):
     with tm.ensure_clean(ext) as f:
         with pytest.raises(ValueError, match=msg):
             ExcelWriter(f, engine="xlsxwriter", mode="a")
+
+
+@pytest.mark.parametrize("c_idx_levels", [1, 2])
+@pytest.mark.parametrize("r_idx_levels", [1, 2])
+@pytest.mark.parametrize("head", [True, False, ["col1", "col2"]])
+@pytest.mark.parametrize("ind", [True, False])
+@pytest.mark.parametrize(
+    "data,fmt,out,default_out",
+    [
+        # Format specified should not have any impact on the data
+        # as the type of the data is int
+        (np.random.randint(1, 100), "0%", "General", None),
+        # Format specified should format the data
+        (
+            datetime(2013, 1, 13, 18, 20, 52),
+            "DD/MM/YYYY",
+            "DD/MM/YYYY",
+            "YYYY-MM-DD HH:MM:SS",
+        ),
+        # Format specified should format the data
+        (date(2014, 1, 31), "MMM", "MMM", "YYYY-MM-DD"),
+        # Format specified should not have any impact on the data
+        # as the type of the data is bool
+        (True, "MMM", "General", None),
+    ],
+)
+def test_num_formats(
+    ext, c_idx_levels, r_idx_levels, head, ind, data, fmt, out, default_out
+):
+    # GH 30275
+    # Testing out num_formats with various dtypes
+    df = tm.makeCustomDataframe(
+        6,
+        2,
+        c_idx_names=False,
+        r_idx_names=False,
+        c_idx_nlevels=c_idx_levels,
+        r_idx_nlevels=r_idx_levels,
+        data_gen_f=lambda r, c: data,
+    )
+    format_dict = dict(zip(df.columns, [fmt] * len(df.columns)))
+
+    if c_idx_levels > 1:
```

Comment:
rather than use multiple parametrization, and then separate out the different parameter cases with logic in a single test, create multiple, simpler tests that test the target operation. E.g. here create a test where if the index is a MultiIndex it ensures "Not Implemented" operates.  Each tests should be as minimal as possible.